cxx/gendb.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#include "gendb.hh"
+
+#include <map>
+#include <random>
+#include <string>
+
+#include "distributions/crp.hh"
+#include "hirm.hh"
+#include "irm.hh"
+#include "observations.hh"
+#include "pclean/schema.hh"
+#include "pclean/schema_helper.hh"
+
+GenDB::GenDB(std::mt19937* prng, const PCleanSchema& schema_,
+             bool _only_final_emissions, bool _record_class_is_clean)
+    : schema(schema_),
+      schema_helper(schema_, _only_final_emissions, _record_class_is_clean) {
+  std::map<std::string, std::vector<std::string>>
+      annotated_domains_for_relation;
+  T_schema hirm_schema =
+      schema_helper.make_hirm_schema(&annotated_domains_for_relation);
+  hirm = new HIRM(hirm_schema, prng);
+
+  for (const auto& [class_name, unused_class] : schema.classes) {
+    domain_crps[class_name] = CRP();
+  }
+}
+
+void GenDB::incorporate(
+    std::mt19937* prng,
+    const std::pair<int, std::map<std::string, ObservationVariant>>& row) {
+  int id = row.first;
+
+  // Maps a query relation name to an observed value.
+  std::map<std::string, ObservationVariant> vals = row.second;
+
+  // Loop over the values in the row, and their query relation names.
+  for (const auto& [query_rel, val] : vals) {
+    // Sample a set of items to be incorporated into the query relation.
+    const std::vector<std::string>& class_path =
+        schema.query.fields.at(query_rel).class_path;
+    std::vector<int> items =
+        sample_entities_relation(prng, schema.query.record_class,
+                                 class_path.cbegin(), class_path.cend(), id);
+
+    // Incorporate the items/value into the query relation.
+    incorporate_query_relation(prng, query_rel, items, val);
```

Comment:
What guarantees that query_rel is a noisy relation here?

Reply:
Updated -- query_rel no longer has to be a noisy relation.